jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,460 @@
+                    return true;
+            }
+        }
+
+        /**
+         * @return false if the pooled object should be disposed of.
+         */
+        public boolean tryRelease()
+        {
+            int newMultiplexingCount;
+            int usageCount;
+            while (true)
+            {
+                long encoded = state.get();
+                usageCount = AtomicBiInteger.getHi(encoded);
+                boolean closed = usageCount < 0;
+                if (closed)
+                    return false;
+
+                newMultiplexingCount = AtomicBiInteger.getLo(encoded) - 1;
+                if (newMultiplexingCount < 0)
+                {
+                    if (LOGGER.isDebugEnabled())
+                        LOGGER.debug("try release failed on negative multiplexing count : {}", newMultiplexingCount);
+                    return false;
+                }
+
+                if (state.compareAndSet(encoded, usageCount, newMultiplexingCount))
+                    break;
+            }
+
+            int currentMaxUsageCount = maxUsageCount;
+            boolean overUsed = currentMaxUsageCount > 0 && usageCount >= currentMaxUsageCount;
+            return !(overUsed && newMultiplexingCount == 0);
+        }
+
+        /**
+         * @return true if the pooled object was marked as removed.
+         */
+        public boolean tryRemove()
+        {
+            boolean removed;
+            int lastUsageCount = state.getHi();
+            while (true)
+            {
+                long encoded = state.get();
+                int usageCount = AtomicBiInteger.getHi(encoded);
+                int currentMaxUsageCount = maxUsageCount;
+                if (currentMaxUsageCount > 0 && usageCount >= currentMaxUsageCount)
+                {
+                    removed = state.compareAndSetHi(usageCount, -1);
+                }
+                else
+                {
+                    removed = state.compareAndSet(usageCount, -1, 1, 0);
+                    if (!removed)
+                        removed = state.compareAndSet(usageCount, -1, 0, 0);
+                }
+                if (removed || lastUsageCount == usageCount)
+                    break;
+                lastUsageCount = state.getHi();
+            }
+            return removed;
+        }
+
+        public boolean isClosed()
+        {
+            return state.getHi() < 0;
+        }
+
+        public boolean isIdle()
+        {
+            long encoded = state.get();
+            if (AtomicBiInteger.getHi(encoded) < 0)
+                return false;
+            return AtomicBiInteger.getLo(encoded) == 0;
+        }
+
+        public boolean isInUse()
+        {
+            long encoded = state.get();
+            if (AtomicBiInteger.getHi(encoded) < 0)
+                return false;
+            return AtomicBiInteger.getLo(encoded) > 0;
+        }
+
+        public int getUsageCount()
+        {
+            return state.getHi();
```

Comment:
This would return -1, do we want that?

Reply:
No, we don't. This should never return a value < 0. Let me fix this.